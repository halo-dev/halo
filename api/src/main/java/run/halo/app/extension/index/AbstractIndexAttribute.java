package run.halo.app.extension.index;

import lombok.EqualsAndHashCode;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;
import run.halo.app.extension.GVK;

@EqualsAndHashCode
public abstract class AbstractIndexAttribute<E extends Extension> implements IndexAttribute {
    private final Class<E> objectType;

    /**
     * Creates a new {@link AbstractIndexAttribute} for the given object type.
     *
     * @param objectType must not be {@literal null}.
     */
    public AbstractIndexAttribute(Class<E> objectType) {
        Assert.notNull(objectType, "Object type must not be null");
        Assert.state(isValidExtension(objectType),
            "Invalid extension type, make sure you have annotated it with @" + GVK.class
                .getSimpleName());
        this.objectType = objectType;
    }

    @Override
    public Class<E> getObjectType() {
        return this.objectType;
    }

    boolean isValidExtension(Class<? extends Extension> type) {
        return type.getAnnotation(GVK.class) != null;
    }
}
