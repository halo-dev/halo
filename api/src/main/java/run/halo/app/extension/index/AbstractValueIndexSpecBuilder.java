package run.halo.app.extension.index;

import org.springframework.util.Assert;
import run.halo.app.extension.Extension;

/**
 * Abstract base implementation of {@link IndexSpecBuilder} for value indexes.
 *
 * @param <E> the type of extension
 * @param <K> the type of key
 * @param <B> the type of builder
 * @author johnniang
 * @since 2.22.0
 */
abstract class AbstractValueIndexSpecBuilder<
    E extends Extension,
    K extends Comparable<K>,
    B extends IndexSpecBuilder<E, K, B>
    > implements IndexSpecBuilder<E, K, B> {

    protected final String name;

    protected final Class<K> keyType;

    protected boolean unique = false;

    protected boolean nullable = true;

    protected AbstractValueIndexSpecBuilder(String name, Class<K> keyType) {
        Assert.hasText(name, "Index name must not be blank");
        Assert.notNull(keyType, "Key type must not be null");
        this.name = name;
        this.keyType = keyType;
    }

    public B unique(boolean unique) {
        this.unique = unique;
        return (B) this;
    }

    public B nullable(boolean nullable) {
        this.nullable = nullable;
        return (B) this;
    }

}
