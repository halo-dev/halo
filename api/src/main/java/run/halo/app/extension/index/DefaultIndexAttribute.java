package run.halo.app.extension.index;

import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;

@Deprecated(forRemoval = true, since = "2.22.0")
class DefaultIndexAttribute<E extends Extension, K extends Comparable<K>>
    implements IndexAttribute<E, K> {

    private final Class<E> objectType;

    private final Class<K> keyType;

    private final Function<E, Set<K>> valuesFunc;

    public DefaultIndexAttribute(Function<E, Set<K>> valuesFunc, Class<E> objectType,
        Class<K> keyType) {
        Assert.notNull(valuesFunc, "Values function must not be null");
        Assert.notNull(objectType, "Cannot resolve object type");
        Assert.notNull(keyType, "Cannot resolve key type");
        this.valuesFunc = valuesFunc;
        this.objectType = objectType;
        this.keyType = keyType;
    }

    @Override
    public Class<E> getObjectType() {
        return objectType;
    }

    @Override
    public Class<K> getKeyType() {
        return keyType;
    }

    @Override
    public Set<K> getValues(E e) {
        if (!checkType(e)) {
            throw new IllegalArgumentException("Object type does not match");
        }
        return Optional.ofNullable(this.valuesFunc.apply(e)).orElse(Set.of());
    }

    private boolean checkType(Extension object) {
        return getObjectType().isInstance(object);
    }
}
