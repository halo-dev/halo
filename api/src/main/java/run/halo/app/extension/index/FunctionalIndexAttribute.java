package run.halo.app.extension.index;

import java.util.Set;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import org.springframework.lang.Nullable;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;

@EqualsAndHashCode(callSuper = true)
public class FunctionalIndexAttribute<E extends Extension>
    extends AbstractIndexAttribute<E> {

    @EqualsAndHashCode.Exclude
    private final Function<E, String> valueFunc;

    /**
     * Creates a new {@link FunctionalIndexAttribute} for the given object type and value function.
     *
     * @param objectType must not be {@literal null}.
     * @param valueFunc value function must not be {@literal null}.
     */
    public FunctionalIndexAttribute(Class<E> objectType,
        Function<E, String> valueFunc) {
        super(objectType);
        Assert.notNull(valueFunc, "Value function must not be null");
        this.valueFunc = valueFunc;
    }

    @Override
    public Set<String> getValues(Extension object) {
        var value = getValue(object);
        return value == null ? Set.of() : Set.of(value);
    }

    /**
     * Gets the value for the given object.
     *
     * @param object the object to get the value for.
     * @return returns the value for the given object.
     */
    @Nullable
    public String getValue(Extension object) {
        if (getObjectType().isInstance(object)) {
            return valueFunc.apply(getObjectType().cast(object));
        }
        throw new IllegalArgumentException("Object type does not match");
    }
}
