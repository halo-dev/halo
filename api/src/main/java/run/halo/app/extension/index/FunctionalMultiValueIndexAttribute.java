package run.halo.app.extension.index;

import java.util.Set;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import org.springframework.util.Assert;
import run.halo.app.extension.Extension;
import run.halo.app.extension.Unstructured;

@EqualsAndHashCode(callSuper = true)
public class FunctionalMultiValueIndexAttribute<E extends Extension>
    extends AbstractIndexAttribute<E> {

    @EqualsAndHashCode.Exclude
    private final Function<E, Set<String>> valueFunc;

    /**
     * Creates a new {@link FunctionalIndexAttribute} for the given object type and value function.
     *
     * @param objectType object type must not be {@literal null}.
     * @param valueFunc value function must not be {@literal null}.
     */
    public FunctionalMultiValueIndexAttribute(Class<E> objectType,
        Function<E, Set<String>> valueFunc) {
        super(objectType);
        Assert.notNull(valueFunc, "Value function must not be null");
        this.valueFunc = valueFunc;
    }

    @Override
    public Set<String> getValues(Extension object) {
        if (object instanceof Unstructured unstructured) {
            var ext = Unstructured.OBJECT_MAPPER.convertValue(unstructured, getObjectType());
            return getNonNullValues(ext);
        }
        if (getObjectType().isInstance(object)) {
            return getNonNullValues(getObjectType().cast(object));
        }
        throw new IllegalArgumentException("Object type does not match");
    }

    private Set<String> getNonNullValues(E object) {
        var values = valueFunc.apply(object);
        return values == null ? Set.of() : values;
    }
}
