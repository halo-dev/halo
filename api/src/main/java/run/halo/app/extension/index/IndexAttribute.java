package run.halo.app.extension.index;

import java.util.Set;
import run.halo.app.extension.Extension;

public interface IndexAttribute {

    /**
     * Specify this class is belonged to which extension.
     *
     * @return the extension class.
     */
    Class<? extends Extension> getObjectType();

    /**
     * Get the value of the attribute.
     *
     * @param object the object to get value from.
     * @param <E> the type of the object.
     * @return the value of the attribute must not be null.
     */
    <E extends Extension> Set<String> getValues(E object);
}
