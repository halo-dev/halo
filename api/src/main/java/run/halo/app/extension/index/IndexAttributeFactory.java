package run.halo.app.extension.index;

import java.util.Set;
import java.util.function.Function;
import lombok.experimental.UtilityClass;
import run.halo.app.extension.Extension;

@UtilityClass
public class IndexAttributeFactory {

    public static <E extends Extension> IndexAttribute simpleAttribute(Class<E> type,
        Function<E, String> valueFunc) {
        return new FunctionalIndexAttribute<>(type, valueFunc);
    }

    public static <E extends Extension> IndexAttribute multiValueAttribute(Class<E> type,
        Function<E, Set<String>> valueFunc) {
        return new FunctionalMultiValueIndexAttribute<>(type, valueFunc);
    }
}
