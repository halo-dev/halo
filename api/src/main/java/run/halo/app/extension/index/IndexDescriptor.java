package run.halo.app.extension.index;

import lombok.Data;
import lombok.ToString;

@Data
@ToString(callSuper = true)
public class IndexDescriptor {

    private final IndexSpec spec;

    /**
     * Record whether the index is ready, managed by {@code IndexBuilder}.
     */
    private boolean ready;

    public IndexDescriptor(IndexSpec spec) {
        this.spec = spec;
    }
}
