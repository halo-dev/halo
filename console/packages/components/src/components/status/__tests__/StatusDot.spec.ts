import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";
import { VStatusDot } from "../index";

describe("StatusDot", () => {
  it("should render", () => {
    expect(mount(VStatusDot)).toBeDefined();
  });

  it("should match snapshot", () => {
    const wrapper = mount(VStatusDot);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should work with state prop", () => {
    ["default", "success", "warning", "error"].forEach((state) => {
      const wrapper = mount(VStatusDot, { props: { state } });
      expect(wrapper.classes()).toContain(`status-dot-${state}`);
    });
  });

  it("should work with animate prop", () => {
    const wrapper = mount(VStatusDot, { props: { animate: true } });
    expect(wrapper.classes()).toContain("status-dot-animate");
  });

  it("should work with text prop", () => {
    const wrapper = mount(VStatusDot, { props: { text: "text" } });
    expect(wrapper.find(".status-dot-text").text()).toBe("text");
  });
});
