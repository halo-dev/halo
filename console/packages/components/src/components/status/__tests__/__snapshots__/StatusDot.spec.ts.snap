// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatusDot > should match snapshot 1`] = `
"<div class=\\"status-dot-wrapper status-dot-success\\">
  <div class=\\"status-dot-body\\"><span class=\\"status-dot-inner\\"></span></div>
  <!--v-if-->
</div>"
`;
